core/src/main/scala/org/broadinstitute/dsde/rawls/util/LibraryPermissionsSupport.scala

```diff
@@ -0,0 +1,106 @@
+import org.broadinstitute.dsde.rawls.model.{ErrorReport, WorkspaceAccessLevels, _}
+import spray.http.StatusCodes
+
+import scala.collection.Set
+
+/**
+  * Created by ahaessly on 3/31/17.
+  */
+trait LibraryPermissionsSupport extends RoleSupport {
+  final val publishedFlag = AttributeName.withLibraryNS("published")
+  final val discoverableWSAttribute = AttributeName.withLibraryNS("discoverableByGroups")
+
+  def withLibraryPermissions(ctx: SlickWorkspaceContext,
+                             operations: Seq[AttributeUpdateOperation],
+                             dataAccess: DataAccess,
+                             userInfo: UserInfo,
+                             isCurator: Boolean,
+                             getMaxAccess: (RawlsUser, SlickWorkspaceContext, DataAccess) => ReadAction[WorkspaceAccessLevel])
+                            (op: => ReadWriteAction[Workspace]): ReadWriteAction[Workspace] = {
+    val names = operations.map(attribute => attribute.name)
+    for {
+      canShare <- dataAccess.workspaceQuery.getUserSharePermissions(RawlsUserSubjectId(userInfo.userSubjectId), ctx)
+      hasCatalogOnly <- dataAccess.workspaceQuery.getUserCatalogPermissions(RawlsUserSubjectId(userInfo.userSubjectId), ctx)
+      userLevel <- getMaxAccess(RawlsUser(userInfo), ctx, dataAccess)
+      result <- getPermissionChecker(names, isCurator, canShare, hasCatalogOnly, userLevel).withPermissions(op)
+    } yield result
+  }
+
+  def getPermissionChecker(names: Seq[AttributeName], isCurator: Boolean, canShare: Boolean, hasCatalogOnly: Boolean, userLevel: WorkspaceAccessLevel) = {
+    val hasCatalog = hasCatalogOnly && userLevel >= WorkspaceAccessLevels.Read
+    // need to multiple delete and add ops when changing discoverable attribute
+    names.distinct match {
+      case Seq(`publishedFlag`) => ChangePublishedChecker(isCurator, hasCatalog, userLevel)
+      case Seq(`discoverableWSAttribute`) => ChangeDiscoverabilityChecker(canShare, hasCatalog, userLevel)
+      case x if x.contains(publishedFlag) => throw new RawlsException("Unsupported parameter - can't modify published with other attributes")
+      case x if x.contains(discoverableWSAttribute) => ChangeDiscoverabilityAndMetadataChecker(canShare, hasCatalog, userLevel)
+      case _ => ChangeMetadataChecker(hasCatalog, userLevel)
+    }
+  }
+
+  def withLibraryAttributeNamespaceCheck[T](attributeNames: Iterable[AttributeName])(op: => T): T = {
+    val namespaces = attributeNames.map(_.namespace).toSet
+
+    // only allow library namespace
+    val invalidNamespaces = namespaces -- Set(AttributeName.libraryNamespace)
+    if (invalidNamespaces.isEmpty) op
+    else {
+      val err = ErrorReport(statusCode = StatusCodes.Forbidden, message = s"All attributes must be in the library namespace")
+      throw new RawlsExceptionWithErrorReport(errorReport = err)
+    }
+  }
+}
+
+trait LibraryPermissionChecker {
```

Comment:
I think this would be better expressed [with functions](https://gist.github.com/dvoet/c750a1d7ebe70d8f706752afb59effb4) instead of command objects.